src/library/scala/collection/GenMapLike.scala

```diff
@@ -124,7 +124,11 @@ trait GenMapLike[K, +V, +Repr] extends GenIterableLike[(K, V), Repr] with Equals
       try {
         val checker = new AbstractFunction1[(K, V),Boolean] with Function0[V]{
           override def apply(kv: (K,V)): Boolean = {
-            that.getOrElse(kv._1.asInstanceOf[b], this) == kv._2
+            // Note: uncurry optimizes this to `get.getOrElse(..., this: Function0)`;  there is no extra lambda allocated.
+            val v2 = that.getOrElse(kv._1.asInstanceOf[b], this)
```

Comment:
Actually, we need to call `this.apply()` in the source here. That will trigger the optimization in uncurry: rather than it emitting `() => this.apply()`, it would opt for `this`.

Currently there is a lambda allocation:

```
  // access flags 0x1
  // signature (Lscala/Tuple2<TK;TV;>;)Z
  // declaration: boolean apply(scala.Tuple2<K, V>)
  public apply(Lscala/Tuple2;)Z
    // parameter final  kv
    ALOAD 0
    GETFIELD scala/collection/GenMapLike$$anon$1.x2$1 : Lscala/collection/GenMap;
    ALOAD 1
    INVOKEVIRTUAL scala/Tuple2._1 ()Ljava/lang/Object;
    ALOAD 0
    INVOKEDYNAMIC apply(Lscala/collection/GenMapLike$$anon$1;)Lscala/Function0; [
      // handle kind 0x6 : INVOKESTATIC
      java/lang/invoke/LambdaMetafactory.altMetafactory(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;[Ljava/lang/Object;)Ljava/lang/invoke/CallSite;
      // arguments:
      ()Ljava/lang/Object;,
      // handle kind 0x6 : INVOKESTATIC
      scala/collection/GenMapLike$$anon$1.$anonfun$apply$1(Lscala/collection/GenMapLike$$anon$1;)Lscala/collection/GenMapLike$$anon$1;,
      ()Lscala/collection/GenMapLike$$anon$1;,
      7,
      1,
      scala.Serializable.class,
      1,
      ()Lscala/collection/GenMapLike$$anon$1;
    ]
    INVOKEINTERFACE scala/collection/GenMap.getOrElse (Ljava/lang/Object;Lscala/Function0;)Ljava/lang/Object; (itf)
```

Reply:
Pushed that change:

```
  // access flags 0x1
  // signature (Lscala/Tuple2<TK;TV;>;)Z
  // declaration: boolean apply(scala.Tuple2<K, V>)
  public apply(Lscala/Tuple2;)Z
    // parameter final  kv
    ALOAD 0
    GETFIELD scala/collection/GenMapLike$$anon$1.x2$1 : Lscala/collection/GenMap;
    ALOAD 1
    INVOKEVIRTUAL scala/Tuple2._1 ()Ljava/lang/Object;
    ALOAD 0
    INVOKEINTERFACE scala/collection/GenMap.getOrElse (Ljava/lang/Object;Lscala/Function0;)Ljava/lang/Object; (itf)
    ASTORE 2
    ALOAD 2
    ALOAD 0
    IF_ACMPEQ L0
    ALOAD 2
    ALOAD 1
    INVOKEVIRTUAL scala/Tuple2._2 ()Ljava/lang/Object;
    INVOKESTATIC scala/runtime/BoxesRunTime.equals (Ljava/lang/Object;Ljava/lang/Object;)Z
    IFEQ L0
    ICONST_1
    IRETURN
   L0
    ICONST_0
    IRETURN
    MAXSTACK = 3
    MAXLOCALS = 3
```

Reply:
> Pushed that change

@retronym did you?

Reply:
Oh, it didn't work:

```
git push -f mkeskells HEAD:mike/2.12.x_MapEquals
Enumerating objects: 13, done.
Counting objects: 100% (13/13), done.
Delta compression using up to 8 threads
Compressing objects: 100% (7/7), done.
Writing objects: 100% (7/7), 918 bytes | 918.00 KiB/s, done.
Total 7 (delta 5), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (5/5), completed with 5 local objects.
To github.com:mkeskells/scala.git
 ! [remote rejected]       HEAD -> mike/2.12.x_MapEquals (permission denied)
error: failed to push some refs to 'git@github.com:mkeskells/scala.git'
```

@mkeskells Here's the suggestion: https://github.com/scala/scala/compare/2.12.x...retronym:review/8875?expand=1